### PR TITLE
(contributing) Add guidelines for anchoring and aliasing of pip entries

### LINKED ISSUE #28891: Pip entries
The current guidelines say that `python-*-pip` entries don't get a `python3-*-pip` version of that entry. (Confirmed by @tfoote in https://github.com/ros/rosdistro/pull/28264#issuecomment-784463002)

But also that new pip entries should use the `python3` prefix (https://github.com/ros/rosdistro/pull/28672#pullrequestreview-614757890) This will result in a big mess of `python-*-pip` and `python3-*-pip` entries. This is not straightforward.

I suggest we either allow the creation of `python3` version of current `python-*-pip` entries, so that in the end all entries exist as a `python3` version and the `python` ones get outdated.

Or we should stick to `python-*-pip` as pip targets are the same for python2 and python3.

Let's clear this up and prevent a big mess.

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
As discussed in #28891

Fixes #28891

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -177,6 +177,24 @@ python3-foobar-pip:
       packages: [foobar]
 ```
 
+In constrast to normal python entries, which are often different for python 2 and 3, pip entries for python 2 and 3 are almost always identical.
+Hence no new entry would be needed. Though this would leave us with a mess of `python3-*`, `python-*-pip` and `python3-*-pip` entries.
+To prevent this, the `python3-*-pip` entry should be mapped to the legacy `python-*-pip` entry by using yaml anchors and aliases.
+(Prerabbly this was the other way around. So the `python3-*-pip` entry containing the contents and the anchor and the legacy `python-*-pip` entry being aliased to it.
```
Comment (tfoote) [inline CONTRIBUTING.md:183]:
```suggestion
(Preferably this was the other way around. So the `python3-*-pip` entry containing the contents and the anchor and the legacy `python-*-pip` entry being aliased to it.
```

**Review by tfoote — APPROVED** [review]
Thanks for the followup. I think that this will work. The name is not completely self explanatory but having this document written out hopefully will help people find the full meaning.

One typo to fixup before merge.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -177,6 +177,24 @@ python3-foobar-pip:
       packages: [foobar]
 ```
 
+In constrast to normal python entries, which are often different for python 2 and 3, pip entries for python 2 and 3 are almost always identical.
```
Comment (cottsay) [inline CONTRIBUTING.md:180]:
```suggestion
In contrast to normal python entries, which are often different for python 2 and 3, pip entries for python 2 and 3 are almost always identical.
```

**Review by cottsay — CHANGES_REQUESTED** [review]
Could you add clarification to the document what date should be used in the anchor?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -177,6 +177,24 @@ python3-foobar-pip:
...
+Hence no new entry would be needed. Though this would leave us with a mess of `python3-*`, `python-*-pip` and `python3-*-pip` entries.
+To prevent this, the `python3-*-pip` entry should be mapped to the legacy `python-*-pip` entry by using yaml anchors and aliases.
+(Preferably this was the other way around. So the `python3-*-pip` entry containing the contents and the anchor and the legacy `python-*-pip` entry being aliased to it.
+Though the anchor should be defined before the anchor is used and `python3-` entries come after `python-` entries in alphabetical order.)
+
+For example:
+
+```yaml
+python-foobar-pip: &migrate_eol_2025_04_30_python3_foobar_pip # Anchor
```
Comment (ivanpauno) [inline CONTRIBUTING.md:189]: why the anchor needs to have a date and a `migrate_eol` prefix?
can't the anchor just be the name of the key? e.g. `&python3_foobar_pip`

Comment (MatthijsBurgh) [reply CONTRIBUTING.md:189]: This is suggested by @tfoote in https://github.com/ros/rosdistro/issues/28891

Comment (ivanpauno) [reply CONTRIBUTING.md:189]: Okey, that makes sense.
Information about how to pick the EOL date should be added here though.

**Review by nuclearsandwich — APPROVED** [review]
In general I think we have to be very cautious when using advanced yaml features both because they're not implemented in all parsers (especially in safe modes) and because they aren't clear to casual yaml editors.

It's a bummer that we have to have the anchors before the aliases as it will create more churn when remove the python- key and move the content to the aliased python3- key.

I cannot think of a different way to solve this same problem and ultimately this is an issue of transition which means that we'll eventually be through it.

Well done @MatthijsBurgh on pushing through and finding a solution to a nasty problem.

**Comment by MatthijsBurgh** [comment]
@cottsay @ivanpauno friendly ping ;)

**Comment by MatthijsBurgh** [comment]
@cottsay @ivanpauno ping ;)

**Review by cottsay — APPROVED** [review]

File: CONTRIBUTING.md (hunk 3 above)
Comment (nuclearsandwich) [reply CONTRIBUTING.md:189]: Resolved by ab459b9

**Comment by nuclearsandwich** [comment]
@MatthijsBurgh thanks for iterating with our team to propose a solution. I'm glad we came up with something that while a bit rough, is no rougher than the problem space, and has an explicit expiration date.

**Pull request merged**